meilisearch-auth/src/index_uid.rs

```diff
@@ -0,0 +1,157 @@
+use serde::de::{Error, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::{Debug, Display, Formatter};
+use std::ops::Deref;
+
+#[derive(Debug, Deserialize, Serialize, Hash, PartialEq)]
+pub struct IndexUid(String);
+
+impl IndexUid {
+    pub fn as_str(&self) -> &str {
+        self.deref()
+    }
+}
+
+impl Display for IndexUid {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl From<IndexUid> for String {
+    fn from(uid: IndexUid) -> Self {
+        uid.0
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidIndexUid(pub String);
+
+impl TryFrom<String> for IndexUid {
+    type Error = InvalidIndexUid;
+
+    fn try_from(uid: String) -> Result<Self, Self::Error> {
+        if uid
+            .chars()
+            .all(|x| x.is_ascii_alphanumeric() || x == '-' || x == '_')
+            || !(1..=400).contains(&uid.len())
```

Comment:
This is duplicated from `meilisearch-lib`, inaccessible from `meilisearch-auth`.
The former already has a `IndexUid` struct representing the same concept, with similar implementation.

I would personally advocate for the creation of a `meilisearch-types` crate that would regroup these implementations, and include all that's in `meilisearch-error`. This new crate could be used to implement Rust clients for Meilisearch. It would be a breaking change though, and it doesn't feel right to me to do this in this PR.

Thoughts?

Reply:
agreed, I don't like that this is duplicated here. Having a `meilisearch-type` crate would solve this issue, and maybe some others.